### PR TITLE
NIFI-7051 Protect against empty group membership in ShellUserGroupPro…

### PR DESCRIPTION
**Pull request by bbende**
…vider, and add differentiator to id seeding

Thank you for submitting a contribution to Apache NiFi.

Please provide a short description of the PR here:

#### Description of PR

_Enables X functionality; fixes bug NIFI-YYYY._

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [ ] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `master`)?

- [ ] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [ ] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [ ] Have you written or updated unit tests to verify your changes?
- [ ] Have you verified that the full build is successful on both JDK 8 and JDK 11?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [ ] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [ ] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by bbende** [comment]
Realized that changing the identifier generation would impact users that already have policies created against this provider. Will work on updating the PR to add a new config option to turn on this new behavior.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-shell-authorizer/src/main/java/org/apache/nifi/authorization/NssShellCommands.java [hunk 1]
```diff
@@ -85,6 +85,6 @@ public String getGroupById(String groupId) {
      * @return Shell command string that will exit normally (0) on a suitable system.
      */
     public String getSystemCheck() {
-        return "getent passwd";
+        return "getent passwd | cut -f 1,3,4 -d ':'";
```
Comment (natural) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-shell-authorizer/src/main/java/org/apache/nifi/authorization/NssShellCommands.java]: I'm not certain the system check needs a pipe and a cut, we're using that to determine if the host can provide `getent` and not any of the output.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-shell-authorizer/src/main/java/org/apache/nifi/authorization/ShellUserGroupProvider.java [hunk 2]
```diff
@@ -272,7 +281,7 @@ public void onConfigured(AuthorizerConfigurationContext configurationContext) th
             }catch (final Throwable t) {
                 logger.error("", t);
             }
-        }, fixedDelay, fixedDelay, TimeUnit.SECONDS);
+        }, fixedDelay, fixedDelay, TimeUnit.MILLISECONDS);
```
Comment (natural) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-shell-authorizer/src/main/java/org/apache/nifi/authorization/ShellUserGroupProvider.java]: Why the change in units?

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-shell-authorizer/src/main/java/org/apache/nifi/authorization/ShellUserGroupProvider.java [hunk 3]
```diff
@@ -194,10 +198,12 @@ public UserAndGroups getUserAndGroups(String identity) throws AuthorizationAcces
         logger.debug("Retrieved user {} for identity {}", new Object[]{user, identity});
 
         Set<Group> groups = new HashSet<>();
-
-        for (Group g : getGroups()) {
-            if (user != null && g.getUsers().contains(user.getIdentity())) {
-                groups.add(g);
+        if (user != null) {
+            for (Group g : getGroups()) {
+                if (user != null && g.getUsers().contains(user.getIdentifier())) {
```
Comment (natural) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-shell-authorizer/src/main/java/org/apache/nifi/authorization/ShellUserGroupProvider.java]: I'm not sure the null check needs to be repeated.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-shell-authorizer/src/main/java/org/apache/nifi/authorization/NssShellCommands.java (hunk 1 above)
Comment (bbende) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-shell-authorizer/src/main/java/org/apache/nifi/authorization/NssShellCommands.java]: The problem is the result set can be really really large and the cut reduces the amount of info returned. In my test environment, without the cut, it was actually taking longer than the 30 second timeout on the Shell runner.

I was actually wondering if we do something that would be quicker, like can we assume every system with NSS has a user named rood and call "getent passwd root" to just look up one user instead of thousands?

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-shell-authorizer/src/main/java/org/apache/nifi/authorization/ShellUserGroupProvider.java (hunk 2 above)
Comment (bbende) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-shell-authorizer/src/main/java/org/apache/nifi/authorization/ShellUserGroupProvider.java]: The fixedDelay variable is being stored as milliseconds:

https://github.com/apache/nifi/blob/master/nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-shell-authorizer/src/main/java/org/apache/nifi/authorization/ShellUserGroupProvider.java#L322

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-shell-authorizer/src/main/java/org/apache/nifi/authorization/ShellUserGroupProvider.java (hunk 3 above)
Comment (bbende) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-shell-authorizer/src/main/java/org/apache/nifi/authorization/ShellUserGroupProvider.java]: Good point, I actually didn't notice the inner null check there.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-shell-authorizer/src/main/java/org/apache/nifi/authorization/NssShellCommands.java (hunk 1 above)
Comment (natural) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-shell-authorizer/src/main/java/org/apache/nifi/authorization/NssShellCommands.java]: Maybe `which getent` is sufficient?  The only required behavior is for the check command to return non-zero.

Comment (bbende) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-shell-authorizer/src/main/java/org/apache/nifi/authorization/NssShellCommands.java]: Let me try that quickly and if it works then I'll update the PR with that, if not I will leave it as is.

Comment (bbende) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-shell-authorizer/src/main/java/org/apache/nifi/authorization/NssShellCommands.java]: Ok I tested it with `getent --version` and that worked so that should be a quicker alternative.

**Comment by bbende** [comment]
I've made the final updates and I believe this is good to go.

**Comment by joewitt** [comment]
+1
